### PR TITLE
CollabFilterClassifier

### PR DESCRIPTION
**Pull request by lrajlich**
Made a stab at the classifier challenge jeremy suggested in lecture on monday night. My understanding is that for a classifier the differences we need are `F.cross_entropy` instead of `F.mse` for calculating loss and to use `F.sigmoid` directly to calculate the activations. 

Looking at the `EmbeddingDotBias`, the activations look like they should evaluate to sigmoid without any code change if your input is all 0 or 1.

Ran through movie lens database and got a loss of .2. 
https://github.com/lrajlich/fastai/blob/CollabFilterClassifier-notebook/courses/dl1/lesson5-movielens.ipynb

Let me know if this is correct or point me in the right direction if I'm off.

If this is right, don't merge right away, I want to validate the input that the ratings are all 0 or 1 (otherwise I assume the model won't work properly).

### CODE REVIEW

**Comment by yanneta** [comment]
You can look at how this is handled in convolutional networks. We just pass an argument instead of writing a new class.

**Comment by lrajlich** [comment]
@yanneta sounds good, I assume you mean to accept an `is_reg` parameter similar to https://github.com/fastai/fastai/blob/master/fastai/conv_learner.py#L23., which I'll have default as `true` to preserve existing usage. Let me know otherwise. I'll also modify the model activation function so that the model will work with classes other than just `0` and `1`, similarly to ConvLearner.

**Comment by yanneta** [comment]
Yes!

**Comment by jph00** [comment]
@lrajlich are you planning to update this PR with @yanneta 's suggestion?

**Comment by lrajlich** [comment]
yes, I haven't had a chance to work on it until today, I'm running into some issues which I'm still resolving but I should have something in a day or two.

**Comment by lrajlich** [comment]
I had a chance to update the PR. The amount of places where `is_reg` has to be passed around is too high in my opinion, which can be resolved by modifying `PassThruDataset`. A more long winded way to explain it is that in ConvLearner is_reg is determined via `data.test_ds.is_reg`; however, with `CollabFilterDataset` the test_ds is a `PassThruDataset` and thus does not have the `is_reg` property on it, so I've had to pass it around. Modifying the `PassThruDataset` to try to make it inherit `BaseDataset` will resolve this issue, which I don't think will be too crazy.

More pertinently, I was seeing issues with `F.cross_entropy`... it expects a rank 2 or rank 4 tensor - the rank 2 seems obvious to me (it's a probability for each class, across each item in the test dataset), rank 4, not so much. The model in this case spits out a rank 1 tensor. It appears that ConvLearner uses `F.binary_cross_entropy` which works with arbitrary sized input and I was able to have the code run here using that. HOWEVER, in ConvLearner, the model is generating a prediction for each class, whereas the model in this case is only generating a single prediction... In order to appropriately build a classifier here, I believe the model would need to be changed to generate a vector of predictions for each class, is this correct? It's not immediately clear to me how to change the model for this. Any guidance on this would be appreciated and I'm interested to tackle the problem.

Test notebook is here: https://github.com/lrajlich/fastai/blob/CollabFilterLearner-is_reg-notebook/courses/dl1/lesson5-movielens_copy.ipynb

**Comment by jph00** [comment]
I agree - that's a lot of places! Any thoughts on how to simplify?

And yes, we shouldn't use binary x-entropy here. We need to use softmax not sigmoid, and then categorical cross-entropy. We do use that in ConvLearner for single label multiclass classification, so you should be able to use a similar approach here. That will be a good exercise :) Feel free to ask on the forum if you get stuck.

**Comment by jph00** [comment]
@yanneta did you have any thoughts to simplify this?

**Comment by lrajlich** [comment]
Sorry, I should have responded earlier... I thought I'd have time to work on this last week but didn't. Basically I can simplify the `is_reg` parameter passing by modifying PassThruDataset have the `is_reg` property, eg, inherit `BaseDataset`. I'm not quite as sure how to address the tensor dimension issue with `F.cross_entropy` (which calls logsoftmax, then nllloss -- this appears to be the same as what ConvnetBuilder does when `is_multi` is false) - Intuitively it seems that the forward method in EmbeddingDotBias class needs to return a higher dimensional tensor than what it's currently returning... I'll look at the lesson6-rnn notebook and see if I can apply a comparable method as the flattening done there.

**Comment by lrajlich** [comment]
I'm going to close this for now (it's been hanging out for a month) and touch on it in a couple of weeks when I get a chance to look into the tensor dimensionality issue with pytorch.
